File: jiahao/dijkstra.py
# dijkstra_baseline_runner.py
import csv
import os
import time
import json
import math
import heapq
from typing import Tuple, List, Optional

MAP_PATH = "./dataset/boston/Boston_0_1024.map"
SAMPLES_CSV_PATH = "./dataset/boston/Boston_0_1024_samples.csv"
OUTPUT_CSV_PATH = "./dataset/boston/Boston_0_1024_results_dijkstra.csv"

# ---------------------------------------------
def load_map(path: str) -> List[List[str]]:
    with open(path, 'r') as f:
        f.readline()
        H = int(f.readline().split()[1])
        W = int(f.readline().split()[1])
        f.readline()
        grid = [list(f.readline().rstrip()) for _ in range(H)]
    return grid

# ---------------------------------------------
def check_collision(pos: Tuple[int, int], grid: List[List[str]], radius: int) -> bool:
    x0, y0 = pos
    H, W = len(grid), len(grid[0])
    for dy in range(-radius, radius + 1):
        for dx in range(-radius, radius + 1):
            x, y = x0 + dx, y0 + dy
            if not (0 <= x < W and 0 <= y < H):
                return False
            if grid[y][x] == '@':
                return False
    return True

# ---------------------------------------------
def get_neighbors(pos: Tuple[int, int], grid: List[List[str]]) -> List[Tuple[int, int]]:
    x, y = pos
    H, W = len(grid), len(grid[0])
    nbrs = []
    for dx, dy in [(0,1),(1,0),(0,-1),(-1,0),(1,1),(-1,-1),(1,-1),(-1,1)]:
        nx, ny = x + dx, y + dy
        if 0 <= nx < W and 0 <= ny < H:
            if abs(dx)==1 and abs(dy)==1 and (grid[y][x+dx]=='@' or grid[y+dy][x]=='@'):
                continue  # corner cut
            nbrs.append((nx, ny))
    return nbrs

# ---------------------------------------------
def dijkstra(grid: List[List[str]], start: Tuple[int,int], goal: Tuple[int,int], radius: int) -> Optional[List[Tuple[int,int]]]:
    open_list = []
    heapq.heappush(open_list, (0.0, start, None))
    came_from = {}
    cost_so_far = {start: 0.0}

    while open_list:
        g, current, parent = heapq.heappop(open_list)

        if current in came_from:
            continue
        came_from[current] = parent

        if current == goal:
            path = []
            while current:
                path.append(current)
                current = came_from[current]
            return path[::-1]

        for neighbor in get_neighbors(current, grid):
            if not check_collision(neighbor, grid, radius):
                continue
            new_cost = cost_so_far[current] + math.hypot(neighbor[0]-current[0], neighbor[1]-current[1])
            if neighbor not in cost_so_far or new_cost < cost_so_far[neighbor]:
                cost_so_far[neighbor] = new_cost
                heapq.heappush(open_list, (new_cost, neighbor, current))

    return None

# ---------------------------------------------
def run():
    grid = load_map(MAP_PATH)

    with open(SAMPLES_CSV_PATH, newline='') as f:
        reader = csv.DictReader(f)
        samples = list(reader)

    fieldnames = reader.fieldnames + ['algo', 'time_ms', 'path']

    with open(OUTPUT_CSV_PATH, 'w', newline='') as outf:
        writer = csv.DictWriter(outf, fieldnames=fieldnames)
        writer.writeheader()

        for row in samples:
            sx = int(row['start_x'])
            sy = int(row['start_y'])
            gx = int(row['goal_x'])
            gy = int(row['goal_y'])
            radius = int(row['radius'])

            start = (sx, sy)
            goal = (gx, gy)

            t0 = time.time()
            path = dijkstra(grid, start, goal, radius)
            t1 = time.time()
            elapsed_ms = (t1 - t0) * 1000.0

            row['algo'] = 'dijkstra'
            row['time_ms'] = f"{elapsed_ms:.3f}"
            row['path'] = json.dumps(path) if path else "[]"
            writer.writerow(row)

    print(f"Results written to {OUTPUT_CSV_PATH}")

if __name__ == "__main__":
    run()

File: jiahao/dijkstra.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <cmath>
#include <chrono>
#include <string>
#include <iomanip>
#include <nlohmann/json.hpp>

using namespace std;
using json = nlohmann::json;

// Map loading
vector<string> loadMap(const string& path) {
    ifstream in(path);
    string line;
    getline(in, line);  // type
    int H, W;
    in >> line >> H;
    in >> line >> W;
    getline(in, line);  // flush
    getline(in, line);  // "map"

    vector<string> grid(H);
    for (int i = 0; i < H; ++i)
        getline(in, grid[i]);
    return grid;
}

// Collision checking
bool checkCollision(const vector<string>& grid, int x0, int y0, int r) {
    int H = grid.size(), W = grid[0].size();
    for (int dy = -r; dy <= r; ++dy) {
        for (int dx = -r; dx <= r; ++dx) {
            int x = x0 + dx, y = y0 + dy;
            if (x < 0 || x >= W || y < 0 || y >= H)
                return false;
            if (grid[y][x] == '@')
                return false;
        }
    }
    return true;
}

// Neighbor generation
vector<pair<int, int>> getNeighbors(int x, int y, const vector<string>& grid) {
    static const vector<pair<int,int>> deltas = {
        {0,1},{1,0},{0,-1},{-1,0}, {1,1},{-1,-1},{1,-1},{-1,1}
    };
    vector<pair<int, int>> nbrs;
    int H = grid.size(), W = grid[0].size();

    for (auto [dx, dy] : deltas) {
        int nx = x + dx, ny = y + dy;
        if (nx < 0 || nx >= W || ny < 0 || ny >= H) continue;
        if (abs(dx)==1 && abs(dy)==1 &&
            (grid[y][x+dx] == '@' || grid[y+dy][x] == '@')) continue;
        nbrs.emplace_back(nx, ny);
    }
    return nbrs;
}

// Dijkstra algorithm
vector<pair<int,int>> dijkstra(
    const vector<string>& grid,
    pair<int,int> start,
    pair<int,int> goal,
    int radius)
{
    typedef pair<double, pair<int,int>> QItem;
    priority_queue<QItem, vector<QItem>, greater<>> open;
    unordered_map<int64_t, pair<int,int>> came_from;
    unordered_map<int64_t, double> cost_so_far;

    auto key = [](int x, int y){ return (int64_t(y) << 32) | uint32_t(x); };

    open.emplace(0.0, start);
    cost_so_far[key(start.first, start.second)] = 0.0;

    while (!open.empty()) {
        auto [g, cur] = open.top(); open.pop();
        int x = cur.first, y = cur.second;

        if (cur == goal) {
            vector<pair<int,int>> path;
            while (cur != start) {
                path.push_back(cur);
                cur = came_from[key(cur.first, cur.second)];
            }
            path.push_back(start);
            reverse(path.begin(), path.end());
            return path;
        }

        for (auto [nx, ny] : getNeighbors(x, y, grid)) {
            if (!checkCollision(grid, nx, ny, radius)) continue;
            double new_cost = g + hypot(nx - x, ny - y);
            int64_t k = key(nx, ny);
            if (!cost_so_far.count(k) || new_cost < cost_so_far[k]) {
                cost_so_far[k] = new_cost;
                came_from[k] = {x, y};
                open.emplace(new_cost, make_pair(nx, ny));
            }
        }
    }
    return {};  // no path
}

// CSV processing
void run(const string& map_path, const string& samples_csv, const string& out_csv) {
    auto grid = loadMap(map_path);
    ifstream inf(samples_csv);
    ofstream outf(out_csv);
    string line;

    getline(inf, line);
    outf << line << ",algo,time_ms,path\n";

    while (getline(inf, line)) {
        istringstream ss(line);
        string tok;
        vector<string> tokens;
        while (getline(ss, tok, ','))
            tokens.push_back(tok);

        int sx = stoi(tokens[0]);
        int sy = stoi(tokens[1]);
        int gx = stoi(tokens[2]);
        int gy = stoi(tokens[3]);
        int radius = stoi(tokens[4]);

        auto t0 = chrono::high_resolution_clock::now();
        auto path = dijkstra(grid, {sx, sy}, {gx, gy}, radius);
        auto t1 = chrono::high_resolution_clock::now();
        double ms = chrono::duration<double, milli>(t1 - t0).count();

        outf << line << ",dijkstra," << fixed << setprecision(3) << ms << ",\"";
        if (!path.empty()) {
            outf << "[";
            for (size_t i = 0; i < path.size(); ++i) {
                outf << "[" << path[i].first << "," << path[i].second << "]";
                if (i + 1 < path.size()) outf << ",";
            }
            outf << "]";
        } else {
            outf << "[]";
        }
        outf << "\"\n";
    }

    cout << "Results written to: " << out_csv << "\n";
}

int main() {
    string map_path = "./dataset/boston/Boston_0_1024.map";
    string samples_csv = "./dataset/boston/Boston_0_1024_samples.csv";
    string out_csv = "./dataset/boston/Boston_0_1024_results_dijkstra.csv";
    run(map_path, samples_csv, out_csv);
    return 0;
}
